Generate MC timestamp

MCruces-fz opened the pull request:
With the intention of adding a function that generates timestamps, this push is presented with the following changes:

- Updated `sensor_utils.py` with the create_timestamp function.
- Added `test_create_timestamp_greater_with_greater_arguments` to `sensor_utils_test.py` with the intention of testing whether with larger parameters said timestamp is greater.
- Updated `buffy.conf`, `buffy.py` and `components.py` to implement the function `create_timestamp`.

Fixes #752

MCruces-fz committed:
Added create_timestamp function and its tests

  - This create_timestamp function allows to generate timestamps for a
given event number and rate.

  - Its test checks if timestamp os greater with greater arguments.

Issue # 752: Generating MC timestamps

andLaing commented on invisible_cities/cities/components_test.py:
```diff
@@ -149,15 +151,15 @@ def dummy_city(files_in, file_out, event_range):
 def test_hits_and_kdst_from_files(ICDATADIR):
     event_number = 1
-    timestamp    = 0.
+    # timestamp    = 0.
```
This test is unrelated so shouldn't be changed

MCruces-fz replied:
Okay, I had assumed that this test would fail and I did not get to check it.

andLaing commented on invisible_cities/cities/components_test.py:
```diff
@@ -149,15 +151,15 @@ def dummy_city(files_in, file_out, event_range):
     assert set(keys) == set(output.keys())
     assert output['event_number']   == event_number
-    assert output['timestamp']      == timestamp
+    # assert output['timestamp']      == timestamp
```
See above

andLaing commented on invisible_cities/cities/components_test.py:
```diff
@@ -191,33 +193,50 @@ def test_copy_mc_info_repeated_event_numbers(ICDATADIR, config_tmpdir):
 def test_mcsensors_from_file_correct_yield(ICDATADIR):
     evt_no         =    0
-    timestamp      =    0
+    # timestamp      =    0
```
Don't leave code commented out. remove the line if you don't need to use it anymore

andLaing commented on invisible_cities/cities/components_test.py:
```diff
@@ -191,33 +193,50 @@ def test_copy_mc_info_repeated_event_numbers(ICDATADIR, config_tmpdir):
 
     assert      first_evt['event_number']                 == evt_no
-    assert      first_evt[   'timestamp']                 == timestamp
+    # assert      first_evt[   'timestamp']                 == timestamp
```
Rather than removing this I'd do a check that it's `>= evt_no / rate`

MCruces-fz replied:
Thank you, good idea

andLaing commented on invisible_cities/cities/components_test.py:
```diff
@@ -191,33 +193,50 @@ def test_copy_mc_info_repeated_event_numbers(ICDATADIR, config_tmpdir):
+    first_evt = next(sns_gen)
+
+    assert first_evt['timestamp'] > timestamp - period
+    assert first_evt['timestamp'] < timestamp + period
```
Is this test to be added in? It seems out of date. I think if you include the rate argument in the source definition it should work.

MCruces-fz replied:
Yes, it is a bad test. It is better to remove it because it is unnecessary.

andLaing commented on invisible_cities/config/buffy.conf:
```diff
@@ -14,4 +14,6 @@ run_number        = -6400
\ No newline at end of file
+trigger_threshold =   0
+
+rate              =   0.5
```
I think we need units here.

`0.5 * hertz` should be valid, for example

andLaing commented on invisible_cities/cities/components.py:
```diff
@@ -341,7 +342,8 @@ def check_lengths(*iterables):
+                        rate      :    float ) -> Generator:
     """
     Loads the nexus MC sensor information into
     a pandas DataFrame using the IC function
```
Add the new parameter to the docstring.

MCruces-fz replied:
Oh, I forgot about that

andLaing commented on invisible_cities/cities/components.py:
```diff
@@ -366,11 +368,10 @@ def mcsensors_from_file(paths     : List[str],
         for evt in mcinfo_io.get_event_numbers_in_file(file_name):
+
+            timestamp = create_timestamp(evt, rate)
+
```
The timestamp that's in the files from data use milliseconds so we should probably be consistent.

Here we'd want to do `timestamp = creat_timestamp(evt, rate) / units.ms`

This solution relies on the correct use of the units in the configuration of the job (I've put a comment to add units to the configuration)

MCruces-fz replied:
Since I'm not used to it, I didn't even think about adding units. I'll keep it in mind next times.

andLaing commented on invisible_cities/detsim/sensor_utils.py:
```diff
@@ -11,6 +11,20 @@
 from .. io      .mcinfo_io import get_sensor_binning
 
 
+def create_timestamp(event_number: int or float,
```
Think the standard way to do this type hint is `Union[int, float]`

andLaing reviewed (changes requested):
Looks good. There's a few stylistic things to sort and the use of the `system_of_units` so that we get the correct final units in the output.

andLaing commented on invisible_cities/cities/components.py:
```diff
@@ -356,6 +358,8 @@ def mcsensors_from_file(paths     : List[str],
     run_number : int
                  Run number for database
+    rate       : float
+                 Rate value in Hertz to generate timestamps
```
So the number (if we use the units correctly) won't be in Hertz. It can be a bit confusing at first but units in this sense convert everything into a base unit. In the case of time we don't define seconds as the base unit (`=1.0`) so Hertz isn't the base unit for rate/frequency either.

The line in the configuration `0.5 * hertz` actually multiplies 0.5 by `1e-9` since the time base unit is `ns`.

I think the least confusing comment might be something like "rate in base unit (ns^-1) but maybe @mmkekic or @jmalbos have an opinion.

MCruces-fz replied:
Right, by reading `system_of_units.py` I have understood the concept about units. Thank you for the explanation

I can wait for other comments and then change the docstring.

andLaing commented on invisible_cities/cities/components_test.py:
```diff
@@ -191,30 +193,31 @@ def test_copy_mc_info_repeated_event_numbers(ICDATADIR, config_tmpdir):
 
     assert      first_evt['event_number']                 == evt_no
-    assert      first_evt[   'timestamp']                 == timestamp
+    assert      first_evt[   'timestamp']                 == evt_no / rate
```
Maybe it's safer as `>=`?

MCruces-fz replied:
True, that was due to lack of attention.

andLaing commented on invisible_cities/detsim/sensor_utils.py:
```diff
@@ -11,6 +11,27 @@
 from .. io      .mcinfo_io import get_sensor_binning
 
 
+def create_timestamp(event_number: Union[int, float],
```
I think you'll need to include `Union` from `typing`, I'm getting a fail on the tests at least.

MCruces-fz replied:
Yes, I need to
`from typing import Union`

Now the pytest does'nt break.

andLaing commented on invisible_cities/detsim/sensor_utils_test.py:
```diff
@@ -97,3 +98,20 @@ def test_pmt_and_sipm_bin_width(full_sim_file):
     assert sipm_binwid == expected_sipmwid
+
+
+def test_create_timestamp_greater_with_greater_arguments():
```
I'd maybe change `arguments` to `eventnumber` both in the name of the test and the docstring

andLaing reviewed (commented):
This is almost ready. Rebase on the master and fix any conflicts then we can see about tidying the history a bit, though probably we can reduce to two commits: the new function and its test added in one and the adjustments to the other functions and tests to accommodate it in another.

MCruces-fz committed:
Adjustements to acommodate the new create_timestamp function

Added rate parameter to buffy function, to mcsensors_from_file
function and rate value to buffy configuration

Updated tests to accept nonzero values in timestamp and new parameters
in modified functions.

andLaing commented on invisible_cities/detsim/sensor_utils.py:
```diff
@@ -34,7 +34,7 @@ def create_timestamp(event_number: Union[int, float],
         timestamp = abs(event_number * period) + np.random.uniform(0, period)
     except ZeroDivisionError:
-        ## Maybe a warning here
+        print(f"WARNING: rate = {rate}")
```
Now that I see it in place I think that putting a warning here (even a proper one) might be unnecessarily verbose. Maybe it's a warning that should go in buffy itself or in the component so it's only sent once per job for the records.

andLaing commented on invisible_cities/cities/components_test.py:
```diff
@@ -29,6 +29,8 @@
 from .. dataflow   import dataflow as fl
 
+from .. detsim.sensor_utils import create_timestamp
+
```
This doesn't get used in this module, I'd remove it.

andLaing reviewed (approved):
This PR adds functionality allowing a user to set a timestamp for Monte Carlo data more similar to that present in data. The function is stand alone so could be used to later change the timestamp in existing MC files if required for a specific analysis.

Tests added and updated where necessary and warning added for unphysical values of the new configuration parameter.

Approved pending integrated testing issue (see #768).
